pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,260 @@
+	predictRouteDestinations := []istiov1alpha3.HTTPRouteDestination{
+		createHTTPRouteDestination(predictDefaultSpec.Hostname, defaultWeight, r.ingressConfig.IngressServiceName),
+	}
+
+	// optionally get a destination for canary predict
+	if kfsvc.Spec.Canary != nil {
+		predictCanarySpec, canaryPredictReason := getPredictStatusConfigurationSpec(kfsvc.Spec.Canary, kfsvc.Status.Canary)
+		if predictCanarySpec != nil {
+			canaryRouteDestination := createHTTPRouteDestination(predictCanarySpec.Hostname, canaryWeight, r.ingressConfig.IngressServiceName)
+			predictRouteDestinations = append(predictRouteDestinations, canaryRouteDestination)
+		} else {
+			return nil, createFailedStatus(canaryPredictReason, "Failed to reconcile canary predictor")
+		}
+	}
+
+	// prepare the predict route
+	predictRoute := istiov1alpha3.HTTPRoute{
+		Match: []istiov1alpha3.HTTPMatchRequest{
+			istiov1alpha3.HTTPMatchRequest{
+				URI: &istiov1alpha1.StringMatch{
+					Prefix: fmt.Sprintf("/v1/models/%s:predict", kfsvc.Name),
+				},
+			},
+		},
+		Route: predictRouteDestinations,
+	}
+	httpRoutes = append(httpRoutes, predictRoute)
+
+	// optionally add the explain route
+	explainRouteDestinations := []istiov1alpha3.HTTPRouteDestination{}
+	if kfsvc.Spec.Default.Explainer != nil {
+		explainDefaultSpec, defaultExplainerReason := getExplainStatusConfigurationSpec(&kfsvc.Spec.Default, kfsvc.Status.Default)
+		if explainDefaultSpec != nil {
+			routeDefaultDestination := createHTTPRouteDestination(explainDefaultSpec.Hostname, defaultWeight, r.ingressConfig.IngressServiceName)
+			explainRouteDestinations = append(explainRouteDestinations, routeDefaultDestination)
+
+			explainCanarySpec, canaryExplainerReason := getExplainStatusConfigurationSpec(kfsvc.Spec.Canary, kfsvc.Status.Canary)
+			if explainCanarySpec != nil {
+				routeCanaryDestination := createHTTPRouteDestination(explainCanarySpec.Hostname, canaryWeight, r.ingressConfig.IngressServiceName)
+				explainRouteDestinations = append(explainRouteDestinations, routeCanaryDestination)
+			} else {
+				return nil, createFailedStatus(canaryExplainerReason, "Failed to reconcile canary explainer")
+			}
+		} else {
+			return nil, createFailedStatus(defaultExplainerReason, "Failed to reconcile default explainer")
+		}
+
+		explainRoute := istiov1alpha3.HTTPRoute{
+			Match: []istiov1alpha3.HTTPMatchRequest{
+				istiov1alpha3.HTTPMatchRequest{
+					URI: &istiov1alpha1.StringMatch{
+						Prefix: fmt.Sprintf("/v1/models/%s:explain", kfsvc.Name),
+					},
+				},
+			},
+			Route: predictRouteDestinations,
+		}
+		httpRoutes = append(httpRoutes, explainRoute)
+	}
+
+	vs := istiov1alpha3.VirtualService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        constants.VirtualServiceName(kfsvc.Name),
+			Namespace:   kfsvc.Namespace,
+			Labels:      kfsvc.Labels,
+			Annotations: kfsvc.Annotations,
+		},
+		Spec: istiov1alpha3.VirtualServiceSpec{
+			Hosts: []string{
+				serviceHostname,
```

Comment:
@yuzisun I can make this "*" then we will be path match only... which means our curl becomes
curl http://`ingress_ip`/v1/models/`model_name`:[predict|explain]

wdyt?

Reply:
actually would have to keep the ns part of the hostname to prevent conflicts across namespaces

Reply:
@ellis-bigelow wdyt? I can reduce the filter so that the model name only appears in the path...

Reply:
I don't think we should be trying to abstract out the model name from the hostname. We should revisit this in the future though. 

Also -- can we make sure we have a catchall routeon /models/model_name (at the entrypoint to the graph) to enable health checks et all.